consensus/istanbul/backend/backend.go

```diff
@@ -874,33 +849,46 @@ func (sb *Backend) retrieveUncachedValidatorConnSet() (map[common.Address]bool,
 	return validatorsSet, nil
 }
 
-func (sb *Backend) sendForwardMsgToProxy(finalDestAddresses []common.Address, ethMsgCode uint64, payload []byte) error {
-	logger := sb.logger.New("func", "sendForwardMsgToProxy")
-	if sb.proxyNode.peer == nil {
-		logger.Warn("No connected proxy for sending a fwd message", "ethMsgCode", ethMsgCode, "finalDestAddreses", common.ConvertToStringSlice(finalDestAddresses))
-		return errNoProxyConnection
+func (sb *Backend) RetrieveValidatorConnSet(retrieveCachedVersion bool) (map[common.Address]bool, error) {
```

Comment:
All these functions seem to exist only to help the `ProxyEngine` in that case, I would define an interface:

```go
type ProxyEngineHelper interface {
  // ... all methods that proxy engine...
}
```
and then pass it to the proxyEngine, instead of sending the backend. That way you decouple them; which makes testing proxyEngine easier. Also you make that connection explicit.

Finally, to avoid adding public methods to `Backend` only to help `ProxyEngine` i would do something like:

```go
type BackendProxyEngineHelper Backend

func (helper *BackendProxyEngineHelper) RetrieveValidatorConnSet(..).. {

}
...
```

and when you create the proxy just do 
```go
sb.proxyEngine = proxy.New((*BackendProxyEngineHelper)(sb))
```

i might have some compilation problems with the pointer and the cast, but the overall schema works.

Again, i would keep the `istanbul.Backend` public interface with only the methods that makes sense for "external" users. In this case, proxy is subyugated to the backend, and it shouldn't make us add public methods just for him

Reply:
I created a `BackendForProxy` and a `BackendForCore` interface here:  https://github.com/celo-org/celo-blockchain/pull/1026/files#diff-d81f86d5bb5bed07a99ff8198470f000

Does that satisfy your concern?  BTW, the only two external users of backend is `proxy` and `core` package, (and eventually the `announce` package).

Reply:
It's better yes. I prefer the suggested approach, but i think it's good enough too. Again, maybe personal taste.

The only thing to consider with this solution, is that user should then use the interfaces as the field member type, and not the actual implementation.

Reply:
I created a proxyEngine, proxiedValidatorEngine objects and BackendForProxyEngine and BackendForProxiedValidatorEngine interfaces that can be used respectively.